src/integrity-checker.js

```diff
@@ -259,28 +320,33 @@ export default class InstallationIntegrityChecker {
       p => !lockfile[p] && (!workspaceLayout || !workspaceLayout.getManifestByPattern(p)),
     );
 
-    const loc = await this._getIntegrityHashLocation();
+    const loc = await this._getIntegrityFileLocation();
     if (missingPatterns.length || !loc.exists) {
       return {
         integrityFileMissing: !loc.exists,
         missingPatterns,
       };
     }
 
-    const actual = await this._generateIntegrityFile(
-      lockfile,
-      patterns,
-      Object.assign({}, flags, {checkFiles: false}), // don't generate files when checking, we check the files below
-      await this._getModuleLocation(),
-    );
+    const actual = await this._generateIntegrityFile(lockfile, patterns, flags, workspaceLayout);
+
     const expected = await this._getIntegrityFile(loc.locationPath);
-    const integrityMatches = await this._compareIntegrityFiles(
-      actual,
-      expected,
-      flags.checkFiles,
-      loc.locationFolder,
-      workspaceLayout,
-    );
+    let integrityMatches = this._compareIntegrityFiles(actual, expected, flags.checkFiles, workspaceLayout);
+    console.log(actual, expected, integrityMatches);
+
+    if (integrityMatches === 'OK') {
+      invariant(expected, "The integrity shouldn't pass without integrity file");
+      for (const modulesFolder of expected.modulesFolders) {
+        console.log(
+          modulesFolder,
+          this.config.lockfileFolder,
+          await fs.exists(path.join(this.config.lockfileFolder, modulesFolder)),
+        );
+        if (!await fs.exists(path.join(this.config.lockfileFolder, modulesFolder))) {
```

Comment:
Nope

  - the `files` entry contains the list of files, but is only there if we use the `--check-files` option. Otherwise it does no check at all.

  - the `modulesFolders` entry is a static list of directory that isn't baked by the filesystem. They are not the actual physical folders, they are the theoretical folders. So when we compare them in `_compareIntegrityFiles`, we don't actually check they exist. That's done here.